block/block.md

```diff
@@ -1 +1,99 @@
+
+#### Building the Block
+
+The block manager of the sequencer nodes performs the following steps to produce a block:
+
+* Call `CreateBlock` using executor
+* Sign the block using `signing key` to generate commitment
+* Call `ApplyBlock` using executor to generate an updated state
+* Save the block, validators, and updated state to local store
+* Add the newly generated block to `pendingBlocks` queue
+* Publish the newly generated block to channels to notify other components of the sequencer node (such as block and header gossip)
+
+### Block Publication to DA Network
+
+The block manager of the sequencer full nodes regularly publishes the produced blocks (that are pending in the `pendingBlocks` queue) to the DA network using the `DABlockTime` configuration parameter defined in the block manager config. In the event of failure to publish the block to the DA network, the manager will perform `maxSubmitAttempts` (default is 30) attempts and an exponential backoff interval between the attempts. The exponential backoff interval starts off at 100 milliseconds and it doubles in the next attempt and capped at `DABlockTime`. A successful publish event leads to the emptying of `pendingBlocks` queue and a failure event leads to proper error reporting.
+
+### Block Retrieval from DA Network
+
+The block manager of the full nodes regularly pulls blocks from the DA network at `DABlockTime` intervals and starts off with a DA height read from the last state stored in the local store or `DAStartHeight` configuration parameter, whichever is the latest. The block manager also actively maintains and increments the `daHeight` counter after every DA pull. The pull happens by making the `RetrieveBlocks(daHeight)` request using the Data Availability Light Client (DALC) retriever, which can return either `Success`, `NotFound`, or `Error`. In the event of an error, a retry logic kicks in after a delay of 100 milliseconds delay between every retry and after 10 retries, an error is logged and the `daHeight` counter is not incremented, which basically results in the intentional stalling of the block retrieval logic. In the block `NotFound` scenario, there is no error as it is acceptable to have no rollup block at every DA height. The retrieval successfully increments the `daHeight` counter in this case. Finally, for the `Success` scenario, first, blocks that are successfully retrieved are marked as hard confirmed and are sent to be applied (or state update). A successful state update triggers fresh DA and block store pulls without respecting the `DABlockTime` and `BlockTime` intervals.
```

Comment:
currently we don't try to re-apply the block upon DA retrieval success. this is not needed because the block is previously applied upon p2p retrieval. note that, if the block is changed, it won't have the same block hash, hence no need for re-apply and simply marking it as hard confirmed should be sufficient?

Reply:
Do we only NOT apply the block if we get it over p2p? Then it is fine. IF we check the block hash before. Do we?  

My question rephrased: What if we don't get it over p2p but only from the DA-Layer? Would we hard confirm it even *before*  applying to the state?

Reply:
yes, if the block is only coming from DA, it is applied. Basically, the logic does make sure to only apply the block once, thats all.

Reply:
Ok, I think what we should do here is define hard confirmed because we think about it probably differently.

Reply:
Follow up: https://github.com/rollkit/rollkit/issues/1240